doc/devel/MEP/MEP28.rst

```diff
@@ -0,0 +1,203 @@
+     and ``np.exp`` for lognormally distributed data)
+  2. ``Axes.boxplot`` will be modified to also accept and naïvely pass them
+     to ``cbook.boxplots_stats`` (Alt: pass the stat function and a dict
+     of its optional parameters).
+  3. Outdated parameters from ``Axes.boxplot`` will be deprecated and
+     later removed.
+
+Implementation
+==============
+
+Passing transform functions to ``cbook.boxplots_stats``
+-------------------------------------------------------
+
+This MEP proposes that two parameters (e.g., ``transform_in`` and
+``transform_out`` be added to the cookbook function that computes the
+statistics for the boxplot function. These will be optional keyword-only
+arguments and can easily be set to ``lambda x: x`` as a no-op when omitted
+by the user. The ``transform_in`` function will be applied to the data
+as the ``boxplot_stats`` function loops through each subset of the data
+passed to it. After the list of statistics dictionaries are computed the
+``transform_out`` function is applied to each value in the dictionaries.
+
+These transformation can then be added to the call signature of
+``Axes.boxplot`` with little impact to that method's complexity. This is
+because they can be directly passed to ``cbook.boxplot_stats``.
+Alternatively, ``Axes.boxplot`` could be modified to accept an optional
+statistical function kwarg and a dictionary of parameters to be direcly
+passed to it.
+
+At this point in the implementation users and external libraries like
+seaborn would have complete control via the ``Axes.boxplot`` method. More
+importantly, at the very least, seaborn would require no changes to its
+API to allow users to take advantage of these new options.
+
+Simplifications to the ``Axes.boxplot`` API and other functions
+---------------------------------------------------------------
+
+Simplifying the boxplot method consists primarily of deprecating and then
+removing the redundant parameters. Optionally, a next step would include
+rectifying minor terminological inconsistencies between ``Axes.boxplot``
+and ``Axes.boxplot``.
+
+The parameters to be deprecated and removed include:
+
+  1. ``usermedians`` - processed by 10 SLOC, 3 ``if`` blocks, a ``for`` loop
+  2. ``conf_intervals`` - handled by 15 SLOC, 6 ``if`` blocks, a ``for`` loop
+  3. ``sym`` - processed by 12 SLOC, 4 ``if`` blocks
+
+Removing the ``sym`` option allows all code in handling the remaining
+styling parameters to be moved to ``Axes.bxp``. This doesn't remove
+any complexity, but does reinforce the single responsibility principle
+among ``Axes.bxp``, ``cbook.boxplot_stats``, and ``Axes.boxplot``.
+
+Additionally, the ``notch`` parameter could be renamed ``shownotches``
+to be consistent with ``Axes.bxp``. This kind of cleanup could be taken
+a step further and the ``whis``, ``bootstrap``, ``autorange`` could
+be rolled into the kwargs passed to the new ``statfxn`` parameter.
```

Comment:
I'll clarify this in the MEP but I'll explain here too. From my POV, there are two options available. The first looks like this:

```python

def boxplot_stats(data, ..., transform_in=None, transform_out=None):
    if transform_in is None:
        transform_in = lambda x: x

    if transform_out is None:
        transform_out = lambda x: x
    
    output = []
    for _d in data:
        d = transform_in(_d)
        stat_dict = do_stats(d)
        for key, value in stat_dict.item():
            if key != 'label':
                stat_dict[key] = transform_out(value)
        output.append(d)
    return output

class Axes(...):
    def boxplot_option1(data, ..., transform_in=None, transform_out=None):
        stats = cbook.boxplot_stats(data, ..., transform_in=transform_in, transform_out=transform_out)
        return self.bxp(stats, ...)

    def boxplot_option2(data, ..., statfxn=None, **statopts):
        if statfxn is None:
            statfxn = boxplot_stats
        stats = statfxn(data, **statopts)
        return self.bxp(stats, ...)
``` 

So in both cases you can do:

``` python
fig, ax1 = plt.subplots()
artists1 = ax1.boxplot_optionX(data, transform_in=np.log, transform_out=np.exp)
```

But Option two lets a user write a completely custom stat function (`my_box_stats`) with fancy BCA confidence intervals and the whiskers set differently depending on some attribute of the data. 

Currently, users can do:

``` python
fig, ax1 = plt.subplots()
my_stats = my_box_stats(data,  bootstrap_method='BCA', whisker_method='dynamic')
ax1.bxp(my_stats)
```

Under Option 2, the user could do (only slightly more convenient):

``` python
fig, ax1 = plt.subplots()
ax1.boxplot(data, statfxn=my_box_stats, bootstrap_method='BCA', whisker_method='dynamic')
```

It's not that compelling of a difference, I admit. *But* Option 2 would let the following be valid in seaborn:

``` python
sns.factorplot(x="day", y="total_bill", hue="sex", data=tips, kind='box',
    palette="PRGn", shownotches=True, statfxn=my_box_stats, 
    bootstrap_method='BCA', whisker_method='dynamic')

```

So a lot of downstream flexibility is gained (in both approaches) with little implementation.